perf/src/sigverify.rs

```diff
@@ -614,13 +614,19 @@ pub fn ed25519_verify(
     // power-of-two number around that accounting for the fact that the CPU
     // may be busy doing other things while being a real validator
     // TODO: dynamically adjust this crossover
-    if valid_packet_count < 64
-        || 100usize
+    let maybe_use_cpu = (valid_packet_count < 64).then_some(true).or_else(|| {
+        100usize
             .wrapping_mul(valid_packet_count)
-            .wrapping_div(total_packet_count)
-            < 90
-    {
-        return ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
+            .checked_div(total_packet_count)
+            .map(|percent_valid| percent_valid < 90)
+    });
+    match maybe_use_cpu {
+        Some(true) => {
+            ed25519_verify_cpu(batches, reject_non_vote, valid_packet_count);
+            return;
+        }
+        Some(false) => (),
+        None => return,
```

Comment:
yeah i assumed as much but didn't want to introduce a panic on that assumption because we aren't sufficiently leveraging type safety to ensure any (un)intentional breakage of that upstream guarantee gets handled here

Reply:
yeah, I think just returning here is better than introducing a panic. I'm guessing we won't hit this case, but not entirely sure about that. Regardless if we hit it or not, it is valid to return immediately if we have no packets to verify.

Reply:
`warpping_div()` would panic if I try to divide by 0[^1].
So, `unwrap()` would not have added a new panic.

It seems nice to handle division by 0 gracefully.
But the `Option<bool>` and the `then_some()`/`or_else()` chain seems to obscure the actual intent.

How about this:
```rust
    let Some(valid_percentage) = 100usize
        .wrapping_mul(valid_packet_count)
        .checked_div(total_packet_count) else {
        return;
    };
    let use_cpu = valid_percentage < 90 || valid_packet_count < 64;
```

[^1]: Here is a simple example:
    ```rust
    fn main() {
        assert_eq!(100i8.wrapping_div(0), 0);
    }
    ```
    ```
    Compiling playground v0.0.1 (/playground)
        Finished dev [unoptimized + debuginfo] target(s) in 0.41s
         Running `target/debug/playground`
    thread 'main' panicked at 'attempt to divide by zero', /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/num/mod.rs:290:5
    note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59ee0662f2a8eff8bb6a9ed33bc5c566#)
    ```
    
    https://play.rust-lang.org/?gist=59ee0662f2a8eff8bb6a9ed33bc5c566

Reply:
>  So, unwrap() would not have added a new panic.

it would have introduced a _misleading_ `unwrap()`. unwrapped `None` vs. attempt to divide by zero. oughta be obvious upon inspection, but still

Reply:
I think, ideally, `unwarp()` should not be used in production code %)
`expect()` is better, as it allows the author to be explicit as to why they think this case should never happen.
With `expect()` the error message could be even better than the one you get from `wrapping_div()`:
```rust
let v = 100usize
    .wrapping_mul(valid_packet_count)
    .checked_div(total_packet_count)
    .expect("`total_packet_count` is non-zero because ...");
```